dpcpp/solver/idr_kernels.dp.cpp

```diff
@@ -55,11 +66,760 @@ namespace dpcpp {
+                    distr(0, 1);
+                auto res = distr(engine);
+
+                work[idx] = res;
+            });
+        });
+    }
+}
+
+
+template <typename ValueType>
+void orthonormalize_subspace_vectors(std::shared_ptr<const DpcppExecutor> exec,
+                                     matrix::Dense<ValueType> *subspace_vectors)
+{
+    orthonormalize_subspace_vectors_kernel<default_block_size>(
+        1, default_block_size, 0, exec->get_queue(),
+        subspace_vectors->get_size()[0], subspace_vectors->get_size()[1],
+        subspace_vectors->get_values(), subspace_vectors->get_stride());
+}
+
+
+template <typename ValueType>
+void solve_lower_triangular(std::shared_ptr<const DpcppExecutor> exec,
+                            const size_type nrhs,
+                            const matrix::Dense<ValueType> *m,
+                            const matrix::Dense<ValueType> *f,
+                            matrix::Dense<ValueType> *c,
+                            const Array<stopping_status> *stop_status)
+{
+    const auto subspace_dim = m->get_size()[0];
+
+    const auto grid_dim = ceildiv(nrhs, default_block_size);
+    solve_lower_triangular_kernel(
+        grid_dim, default_block_size, 0, exec->get_queue(), subspace_dim, nrhs,
+        m->get_const_values(), m->get_stride(), f->get_const_values(),
+        f->get_stride(), c->get_values(), c->get_stride(),
+        stop_status->get_const_data());
+}
+
+
+template <typename ValueType>
+void update_g_and_u(std::shared_ptr<const DpcppExecutor> exec,
+                    const size_type nrhs, const size_type k,
+                    const matrix::Dense<ValueType> *p,
+                    const matrix::Dense<ValueType> *m,
+                    matrix::Dense<ValueType> *alpha,
+                    matrix::Dense<ValueType> *g, matrix::Dense<ValueType> *g_k,
+                    matrix::Dense<ValueType> *u,
+                    const Array<stopping_status> *stop_status)
+{
+    const auto size = g->get_size()[0];
+    const auto p_stride = p->get_stride();
+
+    const dim3 grid_dim(ceildiv(nrhs, default_dot_dim),
+                        exec->get_num_computing_units() * 2);
+    const dim3 block_dim(default_dot_dim, default_dot_dim);
+
+    for (size_type i = 0; i < k; i++) {
+        const auto p_i = p->get_const_values() + i * p_stride;
+        if (nrhs > 1 || is_complex<ValueType>()) {
+            components::fill_array(exec, alpha->get_values(), nrhs,
+                                   zero<ValueType>());
+            multidot_kernel(grid_dim, block_dim, 0, exec->get_queue(), size,
+                            nrhs, p_i, g_k->get_values(), g_k->get_stride(),
+                            alpha->get_values(), stop_status->get_const_data());
+        } else {
+            onemkl::dot(*exec->get_queue(), size, p_i, 1, g_k->get_values(),
+                        g_k->get_stride(), alpha->get_values());
+        }
```

Comment:
Is the oneMKL dot better on one RHS as well like in CUDA? Maybe our implementation is better?

Reply:
I do not test it acutually